rfcs/20200601-tfx-udsl-semantics.md

```diff
@@ -0,0 +1,425 @@
+# Advanced TFX DSL semantics
+
+| Status        | Proposed                                                    |
+| :------------ | :---------------------------------------------------------- |
+| **Author(s)** | Ruoyu Liu (ruoyu@google.com), Konstantin Shtoyk (kostik@google.com), Mitch Trott (trott@google.com), Zhitao Li (zhitaoli@google.com) |
+| **Sponsor**   | Konstantinos Katsiapis (katsiapis@google.com)               |
+| **Updated**   | 2020-04-08                                                  |
+
+## Background
+
+The existing TFX DSL mainly focuses on one-shot pipelines with static execution
+plan. While it is good enough for many use cases, there are some scenarios that
+the current DSL fails to support. Some of those scenarios are becoming more and
+more crucial to modern ML production pipelines. The rest of this section will go
+through some scenarios that this proposal is trying to address.
+
+### Components with different schedules
+
+Until now, all components in a TFX pipeline are expected to be executed the same
+number of times if no failure occurs. For example, in the
+[Chicago Taxi](https://github.com/tensorflow/tfx/tree/v0.21.2/tfx/examples/chicago_taxi_pipeline)
+example, the `ExampleGen` component will be executed no more or no less than the
+`Trainer` component or any other component if every pipeline run finishes
+successfully. This is a strong guarantee but also a strict limitation. The use
+cases shown below are common patterns that require relaxation of this
+limitation.
+
+#### Different input sources
+
+It is common that a ML pipeline uses more than one input sources. It is also
+common that different input sources are generated with different frequencies. In
+this case, it is undesirable to fit the components consuming different data
+sources into the same pipeline with the same schedule since it prevents the
+optimization between data freshness and resources efficiency. On the other hand,
+it is also undesirable to break a logical pipeline into several pieces and run
+them separately since the 'wait for all inputs available' semantics in the
+original pipelines is broken which might result in unintended behaviors such as
+consuming partially ready data. Thus, it is important to be able to author a
+pipeline that consists of components with different schedules and we want such
+experience to be well defined and intuitive.
```

Comment:
Can I describe the use case you mentioned as 'need a way to define trigger condition of a component'?

/cc @1025KB

Reply:
Yup. I would more specifically say 'need a way to trigger components depending on external artifact availability/status' (similar to how components are triggered internally when upstream artifact is ready).

Aka, I want to trigger my an exampleGen whenever the external artifacts that serve as inputs are ready and complete. And this may happen outside of a defined schedule (usually adapts to a schedule but may be delayed). 

CC @newtonle

Reply:
We're thinking of standardizing a 'trigger policy' concept to model scenarios like that. This will also be useful for future features like conditional.